api/src/main/java/org/openmrs/api/impl/TagServiceImpl.java

```diff
@@ -0,0 +1,201 @@
+		return dao.getTagByUuid(uuid);
+	}
+	
+	/**
+	 * @see TagService#saveTag(Tag)
+	 */
+	@Override
+	@Transactional(readOnly = false)
+	public Tag saveTag(Tag tag) {
+		if (getObject(toClass(tag.getObjectType()), tag.getObjectUuid()) == null) {
+			log.warn("Object does not exist in the the database");
+		}
+		return dao.saveTag(tag);
+	}
+	
+	/**
+	 * @see TagService#purgeTag(Tag)
+	 */
+	@Override
+	@Transactional(readOnly = false)
+	public void purgeTag(Tag tag) {
+		dao.deleteTag(tag);
+	}
+	
+	/**
+	 * @see TagService#removeTag(OpenmrsObject, String)
+	 */
+	@Override
+	@Transactional(readOnly = false)
+	public void removeTag(OpenmrsObject openmrsObject, String tag) {
+		Tag tag1 = dao.getTag(openmrsObject, tag);
+		if (tag1 == null) {
+			log.warn("Tag does not exist");
+		} else {
+			purgeTag(tag1);
+		}
+	}
+	
+	/**
+	 * @see TagService#getAllTags()
+	 */
+	@Override
+	public List<String> getAllTags() {
+		return dao.getAllTags();
+	}
+	
+	/**
+	 * @see TagService#getTag(Integer)
+	 */
+	@Override
+	public Tag getTag(Integer tagId) {
+		return dao.getTag(tagId);
+	}
+	
+	/**
+	 * @see TagService#getTags(String)
+	 */
+	@Override
+	public List<Tag> getTags(String searchPhrase) {
+		return dao.getTags(searchPhrase);
+	}
+	
+	/**
+	 * Retrieves an Object with given object uuid and object type from the Openmrs database.
+	 */
+	private <T extends OpenmrsObject> T getObject(Class<T> objectType, String objectUuid) {
+		Object object;
+		object = dao.getObject(objectType, objectUuid);
+		return (T) object;
+	}
+	
+	/**
+	 * @see TagService#addTag(OpenmrsObject, String)
+	 */
+	@Override
+	@Transactional(readOnly = false)
+	public Tag addTag(OpenmrsObject openmrsObject, String tag) {
+		if (hasTag(openmrsObject, tag)) {
+			log.warn("duplicate Tag for " + openmrsObject);
+		} else {
+			Tag tag1 = new Tag(tag, openmrsObject.getUuid(), openmrsObject.getClass().getName());
+			return saveTag(tag1);
+		}
+		return null;
+	}
+	
+	/**
+	 * Validation Method which returns false if u try to add a duplicate tag to an OpenmrsObject
+	 */
+	public boolean hasTag(OpenmrsObject openmrsObject, String tag) {
+		Tag tag1 = dao.getTag(openmrsObject, tag);
+		return tag1 != null;
+	}
+	
+	/**
+	 * @see TagService#getTags(OpenmrsObject)
+	 */
+	@Override
+	public List<Tag> getTags(OpenmrsObject openmrsObject) {
+		return dao.getTags(openmrsObject);
+	}
+	
+	/**
+	 * @see TagService#getTags(List, List)
+	 */
+	@Override
+	public List<Tag> getTags(List<String> objectTypes, List<String> tags) {
+		if (tags.isEmpty()) {
+			throw new APIException("tags cannot be empty");
+		}
+		return dao.getTags(objectTypes, tags);
+	}
+	
+	/**
+	 * Convenience method to return a Class using a string parameter with OpenmrsClassLoader
+	 */
+	private <T extends OpenmrsObject> Class<T> toClass(String className) {
+		Class<T> name = null;
+		try {
+			name = (Class<T>) OpenmrsClassLoader.getInstance().loadClass(className);
+		}
+		catch (ClassNotFoundException e) {}
```

Comment:
It's bad to swallow the exception, get rid of the variable declaration on line 165 above because it's redundant and replace line 167 above with:

`return (Class<T>) OpenmrsClassLoader.getInstance().loadClass(className);`

You will need to re throw the class not found exception wrapped inside an APIException